doc/cookbook/current-route/CurrentRoute.lhs

```diff
@@ -0,0 +1,391 @@
+        req' = req { Wai.vault = vault' }
+    app req' respond
+```
+
+Using this we make third variant of our application.
+Note that we values might not be in the vault,
+that's small functional regression.
+
+```haskell
+data Routes3 route = Routes3
+    { route31 :: route :- Vault :> "route1" :> Get '[JSON] Int
+    , route32 :: route :- Vault :> "route2" :> Capture "id" Int :> Get '[JSON] Int
+    }
+  deriving (Generic)
+
+routes3 :: Routes3 AsServer
+routes3 = Routes3
+    { route31 = \v -> liftIO $ do
+        let p = fromMaybe "?" $ V.lookup rpiKey v
+        BS8.putStrLn p
+        return (BS8.length p)
+    , route32 = \v i -> liftIO $ do
+        let p = fromMaybe "?" $ V.lookup rpiKey v
+        BS8.putStrLn p
+        return (succ i)
+    }
+
+app3 :: Application
+app3 = middleware $ genericServe routes3
+```
+
+Note: that `vault` can be used to pass information from middlewares to handlers
+and from handlers to middlewares. For example, the authentication can be done
+completely in the middleware, with a user information stored in the vault for
+handlers to use.
+
+## How to get current *route*?
+
+The second part of a question, is how to get current route.
+Something, we can get `route2/:id` out?
+Note that handlers are *anonymous*, in the same sense functions are.
+E.g. to write *recursive* anonymous functions, we can use `fix` combinator.
+We can use something close to that to pass "route into itself",
+using `Servant.API.Generics` we can reduce the boilerplate too.
+
+We start with ordinary looking `Routes4` data structure.
+
+```haskell
+data Routes4 route = Routes4
+    { route41 :: route :- "route1" :> Get '[JSON] Int
+    , route42 :: route :- "route2" :> Capture "id" Int :> Get '[JSON] Int
+    }
+  deriving (Generic)
+```
+
+But instead of making a `Routes4 AsServer` value, we'll use a different *mode*.
+`AsRecServer route` is a handler which takes `route :- api` as a first
+argument.  In this example we use `HasLink'`, but reader is free to use other
+automatic interpretations, e.g. `servant-client` to make a proxy!
+
+```haskell
+data AsRecServer route
+instance GenericMode (AsRecServer route) where
+    type AsRecServer route :- api = (route :- api) -> (AsServer :- api)
+
+routes4 :: Routes4 (AsRecServer (AsLink Link))
+routes4 = Routes4
+    { route41 = \l -> liftIO $ do
+        print l
+        return 42
+    , route42 = \l i -> liftIO $ do
+        print (l i)
+        return i
+    }
+
+app4 :: Application
+app4 = genericRecServe routes4
+```
+
+The usage is very simple, unfortunately the implementation is not.
+
+### Hairy bits
+
+The implementation of `genericRecServe` is intimidating.
+The missing bit is a function `genericHoist`.
+In short, given a function which can convert `modeA :- api` into `modeB :- api` for all `api`,
+`genericHoist` converts `routes modeA` into `routes modeB`.
+Maybe this function should exist in `Servant.API.Generic`?
```

Comment:
Note to yourself or do you really mean to ask this question to the reader, hoping one of them will tell us "sure, please add it"? Is this sentence a thunk for the patch that adds this? :-)

Reply:
I'm not sure. It type is quite generic, so it could be in `Servant.API.Generic`, but I don't know if it belongs there with this name and type-signature. 

I'd rephrase that line in

This function might be added into `Servant.API.Generic` later, if it proves to be *generally* useful.

Reply:
Sounds great.

I'm not fond of the name because it's too easy to mix with our existing generic and/or hoist machinery, whereas it's a new kind of transformation here, would be nice to have a dedicated name.

Reply:
Maybe `routesHoist`? It's a `hoist` like function of something we call `routes`.

Reply:
I don't have a better suggestion, that would do.